### PR TITLE
Lazy load term meta

### PR DESCRIPTION
**Pull request by spacedmonkey**
Make it so that term meta is always lazy loaded. 

Trac ticket: https://core.trac.wordpress.org/ticket/57645

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/taxonomy.php [hunk 1]
```diff
@@ -4003,6 +4019,7 @@ function _pad_term_counts( &$terms, $taxonomy ) {
  *
  * @since 4.6.0
  * @since 6.1.0 This function is no longer marked as "private".
+ * @since 6.3.0 Use wp_lazyload_term_meta.
```
Comment (mukeshpanchal27) [inline src/wp-includes/taxonomy.php]:
```suggestion
 * @since 6.3.0 Use wp_lazyload_term_meta() for lazy-loading of term meta.
```

File: tests/phpunit/tests/term/meta.php [hunk 2]
```diff
@@ -199,7 +197,7 @@ public function test_lazy_load_term_meta_false() {
 		foreach ( $terms as $t ) {
 			add_term_meta( $t, 'foo', 'bar' );
 		}
-
+		$this->reset_lazyload_queue();
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/term/meta.php:199]: This call also reset comment queue `$lazyloader->reset_queue( 'comment' );` Don't think it will impact anywhere.

**Review by mukeshpanchal27 — COMMENTED** [review]
Thanks @spacedmonkey, Overall look good from my side. Left document suggestion and feedback. Thanks!

**Comment by spacedmonkey** [comment]
Do you mind taking a look at this @peterwilsoncc ?

File: src/wp-includes/post.php [hunk 3]
```diff
@@ -7660,42 +7658,17 @@ function wp_queue_posts_for_term_meta_lazyload( $posts ) {
...
 						}
 					}
 				}
 			}
 		}
-
-		if ( $prime_term_ids ) {
-			$prime_term_ids = array_unique( $prime_term_ids );
-			// Do not prime term meta at this point, let the lazy loader take care of that.
-			_prime_term_caches( $prime_term_ids, false );
```
Comment (peterwilsoncc) [inline src/wp-includes/post.php]: I'm unclear why the term priming is removed, are you able to drop an explanation here?

File: src/wp-includes/taxonomy.php [hunk 4]
```diff
@@ -1426,6 +1426,22 @@ function update_termmeta_cache( $term_ids ) {
...
+
+/**
+ * Queues term id for lazy-loading of term meta.
+ *
+ * @since 6.3.0
+ *
+ * @param array $term_ids List of term IDs.
+ */
+function wp_lazyload_term_meta( array $term_ids ) {
+	if ( ! $term_ids ) {
```
Comment (peterwilsoncc) [inline src/wp-includes/taxonomy.php]: Maybe, not a strong feeling.

```suggestion
	if ( empty( $term_ids ) ) {
```

File: tests/phpunit/tests/term/meta.php [hunk 5]
```diff
@@ -16,6 +16,7 @@ class Tests_Term_Meta extends WP_UnitTestCase {
 	public function set_up() {
 		parent::set_up();
 		register_taxonomy( 'wptests_tax', 'post' );
+		$this->reset_lazyload_queue();
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/term/meta.php]: I don't understand why this is needed, the code in the default tearDown should ensure that the queues are empty.

File: tests/phpunit/tests/post/nav-menu.php [hunk 6]
```diff
@@ -383,12 +384,14 @@ public function test_wp_get_nav_menu_items_cache_primes_terms() {
...
 		$queries_made = get_num_queries() - $start_num_queries;
 		$this->assertSame( 6, $queries_made, 'Only does 6 database queries when running wp_get_nav_menu_items.' );
 
-		$args = $action_terms->get_args();
-		$last = end( $args );
-		$this->assertSameSets( $term_ids, $last[1], '_prime_term_caches() was not executed.' );
+		$args       = $action_terms->get_args();
+		$first      = reset( $args );
+		$term_ids[] = $this->menu_id;
+		$this->assertSameSets( $term_ids, $first[1], '_prime_term_caches() was not executed.' );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/nav-menu.php:394]: What's the purpose of this change?

**Review by peterwilsoncc — COMMENTED** [review]
The `Tests_Term_WpGetObjectTerms::test_termmeta_cache_should_not_be_primed_when_update_term_meta_cache_is_false` test is failing due to a difference in the number of queries.

Prior to this change, it would make three database queries, one for each term. With this change it makes fewer database queries.

Otherwise a few notes and questions inline

File: src/wp-includes/post.php (hunk 3 above)
Comment (spacedmonkey) [reply src/wp-includes/post.php]: So before, we were doing a call to `get_term` to check it exists. But it is extremely unlikely the term will not exist, making an extensive call to `get_term` unneeded. This prime of terms was only done because of this. The term does not need to be loaded here. 99% of the time, the term is already in memory anyway, making this call wasteful. To prime term meta, all we need is the id.

File: tests/phpunit/tests/term/meta.php (hunk 2 above)
Comment (spacedmonkey) [reply tests/phpunit/tests/term/meta.php:199]: I don't think this is a problem here.

File: src/wp-includes/post.php (hunk 3 above)
Comment (spacedmonkey) [reply src/wp-includes/post.php]: CC @joemcgill

Comment (spacedmonkey) [reply src/wp-includes/post.php]: This change is somewhat unrelated, I will revert it for now.

File: tests/phpunit/tests/post/nav-menu.php (hunk 6 above)
Comment (spacedmonkey) [reply tests/phpunit/tests/post/nav-menu.php:394]: @peterwilsoncc Now that all term meta is lazy loaded and after [this](https://github.com/WordPress/wordpress-develop/commit/7130af403fff492d4bde3e2fbd08bc3d25150dfb) change, the first element in this array, primes the menu's term meta and all other term ids term meta in one go.

File: tests/phpunit/tests/query/lazyLoadTermMeta.php [hunk 7]
```diff
@@ -15,6 +15,11 @@ class Test_Lazy_Load_Term_Meta extends WP_UnitTestCase {
 	 */
 	protected static $term_ids = array();
 
+	public function set_up() {
+		parent::set_up();
+		$this->reset_lazyload_queue();
+	}
+
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/query/lazyLoadTermMeta.php]:
```suggestion
```

I don't think this is needed with the default reset in tearDown. Correct me if I am wrong.

**Review by peterwilsoncc — COMMENTED** [review]
One last note but otherwise I think this is good for commit.

File: tests/phpunit/tests/query/lazyLoadTermMeta.php [hunk 8]
```diff
@@ -45,6 +45,7 @@ public static function wpSetUpBeforeClass( WP_UnitTest_Factory $factory ) {
 	 * @covers ::wp_queue_posts_for_term_meta_lazyload
 	 */
 	public function test_wp_queue_posts_for_term_meta_lazyload() {
+		$this->reset_lazyload_queue();
```
Comment (spacedmonkey) [inline tests/phpunit/tests/query/lazyLoadTermMeta.php:48]: This is needed for the unit test to pass.

Comment (peterwilsoncc) [reply tests/phpunit/tests/query/lazyLoadTermMeta.php:48]: Odd, I suspect it means that the queue should be reset in the setup rather than tear down. A problem for another day.

File: src/wp-includes/taxonomy.php [hunk 9]
```diff
@@ -1426,6 +1426,22 @@ function update_termmeta_cache( $term_ids ) {
 	return update_meta_cache( 'term', $term_ids );
 }
 
+
+/**
+ * Queues term id for lazy-loading of term meta.
```
Comment (peterwilsoncc) [inline src/wp-includes/taxonomy.php]:
```suggestion
 * Queue term meta for lazy-loading.
```

**Review by peterwilsoncc — COMMENTED** [review]
This LGTM, one dockblock suggestion but I reckon it's commit ready.

**Comment by spacedmonkey** [comment]
Committed in [55671](https://core.trac.wordpress.org/changeset/55671).
